File: Relationship/Aggregation/ex1_employee.go
package main

import "fmt"

type Address struct {
	addressLine string
	city        string
	state       string
}

func NewAddress(addressLine string, city string, state string) Address {
	return Address{
		addressLine: addressLine,
		city:        city,
		state:       state,
	}
}

type Employee struct {
	id      int
	name    string
	address Address
}

/*
*****************************************************************************
	new employee by value
*****************************************************************************
*/
func NewEmployee(id int, name string, address Address) *Employee {
	e := Employee{
		id:      id,
		name:    name,
		address: address,
	}
	fmt.Printf("Check address after NewEmployee %p\n", &e)

	return &e
}

/*
*****************************************************************************
	new employee by pointer
*****************************************************************************
*/
func NewEmployeeByPointer(id int, name string, address Address) *Employee {

	e := new(Employee)
	e.id = id
	e.name = name
	e.address = address
	fmt.Printf("Check address after NewEmployeeByPointer %p\n", &e)

	return e
}

/*
*****************************************************************************
	main
*****************************************************************************
*/
func main() {
	address := NewAddress("Hau Giang", "HCM", "70000")
	fmt.Printf("Check address initialize %p\n", &address)

	/******************************************************************************
		create a object using constructors by value
	******************************************************************************/
	employee := NewEmployee(1, "Hua Van Thong", address)
	fmt.Printf("ID: %d ,\nName: %s, \n AddressLine: %s, Address of employee.address %p \n",
		employee.id,
		employee.name,
		employee.address.addressLine,
		&employee.address)

	/******************************************************************************
		create a object using constructors by pointer
	******************************************************************************/
	employee2 := NewEmployeeByPointer(1, "Hua Van Thong", address)
	fmt.Printf("ID: %d ,\nName: %s, \n AddressLine: %s, Address of employee.address %p\n",
		employee2.id,
		employee2.name,
		employee2.address.addressLine,
		&employee2.address)
}

File: Relationship/Aggregation/ex1_employee.java
package main;

// Step 1: Define class Address
// ====> Kinh nghiệm cho thấy là, 
//       không nên sử dụng private access modifier cho các member trong class này
// Bởi vì, nếu ta muốn access từ class khác, nó được public hoặc không define
class Address {  
    String addressLine;
    String city;
    String state;

    public Address(String addressLine, String city, String state)    
    {    
        this.addressLine    = addressLine;    
        this.city           = city;    
        this.state          = state;    
    }    
}

class Employee  
{    
    private int id;  
    private String name;  
    private Address address;  

    public Employee(int id, String name, Address address)    
    {    
        this.id = id;    
        this.name = name;    
        this.address = address;    
    }    
    
    public void display()    
    {    
        System.out.println("ID: " + this.id);
        System.out.println("Name: " + this.name);
        System.out.println("AddressLine: " + this.address.addressLine); //==> Access member trong class Address
        System.out.println("City: " + this.address.city);
        System.out.println("State: " + this.address.state);
    }    
};   


public class Aggregation {
    public static void main(String []Args) {
        // Create a instance reference to object Address
        Address add = new Address("Hau Giang", "HCM", "70000");

        Employee emp = new Employee(1, "Hua Van Thong", add);
        emp.display();
    }
}  